docs/externs.md

```diff
@@ -0,0 +1,159 @@
+# Externs
+
+Externs in Exo provide a mechanism to interface with external functions and libraries directly from your Exo code. By defining custom extern functions, you can extend the capabilities of Exo and leverage existing code written in other languages like C or C++. Externs can be used as expressions in your code, particularly on the right-hand side (RHS) of assignment and reduction statements.
+
+## Defining Externs in User Code
+
+Extern functions are defined by subclassing the `Extern` class provided by Exo. This allows you to specify how the extern function should behave, including type checking, compilation, and any global code it might require.
+
+### Step-by-Step Guide
+
+#### 1. Import the Extern Class
+
+Before you can define an extern function, you need to import the `Extern` class and the `_EErr` exception from `exo.core.extern`.
+
+```python
+from exo.core.extern import Extern, _EErr
+```
+
+- `Extern`: The base class for creating custom extern functions.
+- `_EErr`: An exception class used for error handling during type checking.
+
+#### 2. Subclass the Extern Class
+
+Create a new class that inherits from `Extern`. This class represents your custom extern function.
+
+```python
+class _Sin(Extern):
+    # Implementation details will go here
+```
+
+#### 3. Implement Required Methods
+
+Your subclass must implement several methods to define the behavior of the extern function.
+
+##### `__init__(self)`
+
+Initialize your extern function with its name.
+
+```python
+def __init__(self):
+    super().__init__("sin")
+```
+
+- `"sin"`: The name of the external function as it will appear in the generated code.
+
+##### `typecheck(self, args)`
+
+Define how the function checks the types of its arguments.
+
+```python
+def typecheck(self, args):
+    if len(args) != 1:
+        raise _EErr(f"expected 1 argument, got {len(args)}")
+
+    arg_type = args[0].type
+    if not arg_type.is_real_scalar():
+        raise _EErr(
+            f"expected argument to be a real scalar value, but got type {arg_type}"
+        )
+    return arg_type
+```
+
+- Checks that there is exactly one argument.
+- Ensures the argument is a real scalar type (e.g., `float`, `double`).
+- Returns the type of the argument as the return type of the function.
+
+##### `compile(self, args, prim_type)`
```

Comment:
Elaborate: this returns a string for passing to `str.format`; also, what are is set of arguments (e.g. `prim_type`) that will get passed to `str.format`? Note in particular that `prim_type` does not use the type names for Exo (e.g. `f32`) ... it could be important for the user to know the naming scheme here if the user is doing something clever and not just pasting the `prim_type` directly into generated C code.